src/main/java/com/mentorship/vineservice/services/imp/OrderServiceImpl.java

```diff
@@ -0,0 +1,140 @@
+
+    private static final String CREATE_ORDER_EVENT_MESSAGE = "New order has been created";
+
+    @Value("${spring.rabbitmq.template.exchange}")
+    private String exchange;
+
+    @Value("${spring.rabbitmq.template.routing-key}")
+    private String routingKey;
+
+    @Override
+    public Long createOrder(OrderDto orderDto) throws VinePermissionException {
+
+        Order order = vineMapper.orderDtoToOrder(orderDto);
+        List<OrderVine> orderVineList = vineMapper.orderVineDtoToOrderVine(orderDto.getVines());
+
+        String homeAddress = orderDto.getDeliveryDetails().getHomeAddress();
+        PostOfficeDto postOffice = orderDto.getDeliveryDetails().getPostOffice();
+
+        validateDuplicationDeliveryDetails(homeAddress, postOffice);
+
+        addPostOfficeIfPresent(postOffice, order);
+
+        addVinesToOrder(order, orderVineList);
+
+        Long createdOrderId = orderRepository.save(order).getId();
+        log.info(String.format("New order created with id '%s", createdOrderId));
+
+        updateSoldVineAndVineAmount(orderVineList);
+
+        publishCreatedOrderEvent(CREATE_ORDER_EVENT_MESSAGE, createdOrderId);
+
+        return createdOrderId;
+    }
+
+
+    private void validateDuplicationDeliveryDetails(String homeAddress, PostOfficeDto postOffice)
+        throws VinePermissionException {
+        if (homeAddress == null && postOffice == null) {
+            throw new VinePermissionException(HttpStatus.BAD_REQUEST,
+                "'homeAddress' and 'postOffice' can not be null simultaneously");
+        }
+
+        if (homeAddress != null && postOffice != null) {
+            throw new VinePermissionException(HttpStatus.BAD_REQUEST,
+                "'homeAddress' and 'postOffice' can not be filled at the same order");
+        }
+    }
+
+    public void addPostOfficeIfPresent(PostOfficeDto postOffice, Order order) {
+        if (postOffice != null) {
+            PostOffice postOfficeObject = postOfficeRepository.findPostOfficeByCityAndOfficeNumber(postOffice.getCity(),
+                    postOffice.getOfficeNumber())
+                .orElseThrow(() -> new NoSuchElementException(
+                    String.format("Post office in %s with number %s not exist", postOffice.getCity(),
+                        postOffice.getOfficeNumber())));
+
+            order.getDeliveryDetails().setPostOfficeId(postOfficeObject.getId());
+            log.info(String.format("Post office id '%s' added to order", postOfficeObject.getId()));
+        }
+    }
+
+    private void addVinesToOrder(Order order, List<OrderVine> orderVines) {
+        order.setVines(new ArrayList<>());
+        orderVines.forEach(orders -> {
+            Vine vineObject = vineService.getVineById(orders.getOrderVineId().getVineId());
+            Integer availableVines = vineObject.getAmount();
+            if (orders.getVineAmount() > availableVines) {
+                try {
+                    throw new VinePermissionException(HttpStatus.BAD_REQUEST,
```

Comment:
Permission exception here is not obvious here, when there is not enough wine.  I think it should be IllegalArgument exception or a custom vine exception

Reply:
Added illegalArgument exeption